### PR TITLE
Adds total discount rules

### PR DESCRIPTION
**Pull request by forellana-digithinkit**
Total rule applicable with/without items, with/without services rules(ex: free shipping) based on total percentage or value discount.

Only one condition supported atm: "Greater Than Total"

If total rule method is set to "None", you can force coupons to only apply when a total is reached without applying any total discount.

feature-total-rule

UPDATE:

so I ended up having to extend erpnext's account controller and added a on_calculate_taxes_and_totals hook. The hook lets us extend the calculate_taxes_and_totals class in the taxes_and_totals controller.

You'll need this PR to make this work: https://github.com/DigiThinkIT/erpnext/pull/37

And changes to Quotation, Sales Order and Sales Invoice here: https://github.com/DigiThinkIT/jhaudio_customizations/pull/1098

### CODE REVIEW

**Comment by Alchez** [comment]
@forellana-digithinkit:

- You're hardcoding the cart and applied quantities for the discount state as 1.
- The 3-column layout makes the table look weird. :sweat_smile:

**Comment by forellana-digithinkit** [comment]
@Alchez check it out. Added an update on the PR description

File: awesome_cart/hooks.py [hunk 1]
```diff
@@ -61,6 +61,8 @@
 	}
 }
 
+on_calculate_taxes_and_totals = "awesome_cart.utils.on_calculate_taxes_and_totals"
```
Comment (vjFaLk) [inline awesome_cart/hooks.py]: I'm assuming you've tested this hook, cause I'm not sure if it *should* be in list if you're iterating through it in the code.

File: awesome_cart/utils.py [hunk 2]
```diff
@@ -214,3 +204,7 @@ def run_hotpatch(*args, **kwargs):
 	if not hasattr(get_item_details.process_args, "__patched"):
 		print(" - Patched: erpnext.stock.get_item_details.process_args -> awesome_cart.utils.erpnext_stock_get_item_details")
 		get_item_details.process_args = fn_wrap(get_item_details.process_args, erpnext_stock_get_item_details)
+
+def on_calculate_taxes_and_totals(doc):
```
Comment (vjFaLk) [inline awesome_cart/utils.py]: Can you confirm if it's okay if you only pass it the "doc" and not also the method. How does frappe.call make calls?

File: awesome_cart/hooks.py (hunk 1 above)
Comment (forellana-digithinkit) [reply awesome_cart/hooks.py]: ye, frappe combines all strings into an array. tested it and it works

File: awesome_cart/utils.py (hunk 2 above)
Comment (forellana-digithinkit) [reply awesome_cart/utils.py]: This is OUR hook, only passes the doc. See: https://github.com/DigiThinkIT/erpnext/pull/37

**Pull request merged**